### ISSUE TITLE
[WINE] System.NotImplementedException on IAudioEndpointVolume.GetChannelCount

### ISSUE THREAD

**Issue by Web-eWorks** [ciribob/DCS-SimpleRadioStandalone#621]
Reviving this issue from #409 as the underlying cause of the issue reported in the comments is fairly clear after a bit of digging. Attempting to run SRS-Client on linux in a WINE 7.2 install with .NET 4.6.2 allows the user interface to fully start up, but the application errors and closes when attempting to preview microphone volume with the following error in the log:

```
2022-08-26 23:00:50.6020 | Ciribob.DCS.SimpleRadio.Standalone.Client.Audio.AudioPreview | Error starting audio Input - Quitting! The method or operation is not implemented. System.NotImplementedException: The method or operation is not implemented.
   at NAudio.CoreAudioApi.Interfaces.IAudioEndpointVolume.GetChannelCount(Int32& pnChannelCount)
   at NAudio.CoreAudioApi.AudioEndpointVolumeChannels..ctor(IAudioEndpointVolume parent) in G:\Dropbox\Dev\DCS\DCS-SRS\NAudio\CoreAudioApi\AudioEndpointVolumeChannels.cs:line 63
   at NAudio.CoreAudioApi.AudioEndpointVolume..ctor(IAudioEndpointVolume realEndpointVolume) in G:\Dropbox\Dev\DCS\DCS-SRS\NAudio\CoreAudioApi\AudioEndpointVolume.cs:line 160
   at NAudio.CoreAudioApi.MMDevice.GetAudioEndpointVolume() in G:\Dropbox\Dev\DCS\DCS-SRS\NAudio\CoreAudioApi\MMDevice.cs:line 90
   at Ciribob.DCS.SimpleRadio.Standalone.Client.Audio.AudioPreview.StartPreview(Boolean windowsN)
```

This is caused by the `IAudioEndpointVolume` method `GetChannelCount` (and more generally, all per-channel audio APIs in IAudioEndpointVolume) in WINE being currently [unimplemented](https://github.com/wine-mirror/wine/blob/master/dlls/mmdevapi/audiovolume.c#L119) (most likely due to lack of common feature support in the underlying ALSA/PulseAudio/Jack platform backends that WASAPI delegates to under WINE).

The channel count information is still available to the application via the audio device MixFormat information, it's simply that NAudio throws an exception when GetChannelCount is unavailable when building an AudioEndpointChannels object.

```
2022-08-26 22:59:03.9807 | Ciribob.DCS.SimpleRadio.Standalone.Client.Singletons.AudioOutputSingleton | Building Microphone Audio Outputs 
2022-08-26 22:59:03.9807 | Ciribob.DCS.SimpleRadio.Standalone.Client.Singletons.AudioOutputSingleton | Mic Audio Output - Saved ID  
2022-08-26 22:59:03.9807 | Ciribob.DCS.SimpleRadio.Standalone.Client.Singletons.AudioOutputSingleton | Audio Output - PulseAudio {0.0.0.00000000}.{FD47D9CC-4218-4135-9CE2-0C195C87405B} CHN:2 Rate:44100
```

If support for setting/querying individual channel volume on an IAudioEndpointVolume is not used/needed in SRS, it may be a simple matter to work around this bug and regain the ability to run SRS in WINE; otherwise the channel-specific methods will have to be implemented on the WINE side beyond the scope of this issue.

Unfortunately, I'm not familiar with C# development, so I cannot provide a proof-of-concept of what a functional workaround would look like to implement.

**Comment by ciribob** [comment]
This comes from the NAudio library - though I do have a fork of NAudio and will have a quick look

Thanks for the detail investigation!

**Comment by Web-eWorks** [comment]
As a quick update, I managed to get SRS working under WINE by commenting out three lines in AudioEndpointVolume; will reply with the actual changed code in a few hours when not on mobile.

**Comment by ciribob** [comment]
Oh thats great news! I dont mess with the volume so can integrate that without any issues. Thanks!

**Comment by ciribob** [comment]
Thanks! This will be part of the next release #622

**Comment by TheZoq2** [comment]
I've flown with this for a few days now and it seems to work really well! I had to install dotnet462 and fix WPF using the instructions here https://www.reddit.com/r/linux4noobs/comments/firqs9/getting_windows_wpf_applications_to_run_with_wine/

My first flight, I kept being disconnected from the server, but after upgrading the game scripts to the latest version, it has been flawless.

Thanks to both of you for putting the effort into this for us linux users :heart:

**Comment by Web-eWorks** [comment]
@TheZoq2 just as an extra FYI to throw into the DCS on Linux guide, the winepulse driver will report an audio MixFormat for all devices based on the default recording device the "test" audio stream it creates happens to get assigned to - if the four-channel PS3 Eye camera happens to be the first in the list, winepulse will report to SRS that all input devices have 4 channels and SRS will silently fail to initialize a correct recording stream (it will log a single `Recording Stopped` in the client log due to attempting to pull 4-channel audio from a 1- or 2-channel device).

I solved this by setting the PS Eye camera's audio profile to Off in `pavucontrol` or `pulsemixer`, but given the utter lack of info about this and the prevalence of the PS3 camera in linux headtracking, figured I'd make a mention of it here for posterity.

![image](https://user-images.githubusercontent.com/4218491/192170663-c57ae721-3d8a-4752-ae9f-a7f29d0cfb26.png)

**Comment by TheZoq2** [comment]
Interesting, would you mind making a PR to update the DCS on linux repo with that info?

**Comment by Web-eWorks** [comment]
I can definitely look into it when I get the time!

**Comment by vsTerminus** [comment]
For anyone stumbling on this thread in 2023, the dotnet version that SRS depends on has been bumped up. You still need the WPF fix mentioned here https://github.com/ciribob/DCS-SimpleRadioStandalone/issues/621#issuecomment-1255474316 but now you want dotnet48. You may also want "corefonts" otherwise the interface text is pretty ugly.

Finally, run SRS from its own prefix. Disable auto-launch in DCS and just manually launch it.

Basically, run these three commands:

    WINEPREFIX=/path/to/srs/prefix winetricks -q dotnet48 corefonts
    WINEPREFIX=/path/to/srs/prefix wine reg ADD 'HKCU\Software\Wine\DllOverrides' '/f' '/v' 'd3d9' '/t' 'REG_SZ' '/d' 'native'
    WINEPREFIX=/path/to/srs/prefix wine SR-ClientRadio.exe
    
That worked for me on Arch Linux with wine-staging 8.18

**Comment by NorseManGef** [comment]
> For anyone stumbling on this thread in 2023, the dotnet version that SRS depends on has been bumped up. You still need the WPF fix mentioned here [#621 (comment)](https://github.com/ciribob/DCS-SimpleRadioStandalone/issues/621#issuecomment-1255474316) but now you want dotnet48. You may also want "corefonts" otherwise the interface text is pretty ugly.
> 
> Finally, run SRS from its own prefix. Disable auto-launch in DCS and just manually launch it.
> 
> Basically, run these three commands:
> 
> ```
> WINEPREFIX=/path/to/srs/prefix winetricks -q dotnet48 corefonts
> WINEPREFIX=/path/to/srs/prefix wine reg ADD 'HKCU\Software\Wine\DllOverrides' '/f' '/v' 'd3d9' '/t' 'REG_SZ' '/d' 'native'
> WINEPREFIX=/path/to/srs/prefix wine SR-ClientRadio.exe
> ```
> 
> That worked for me on Arch Linux with wine-staging 8.18

How would I actually go about connecting SRS to DCS while in a separate prefix? I have the SR-ClientRadio.exe running, but how do you link it to DCS?

**Comment by TheZoq2** [comment]
SRS communicates with the game via sockets, so it doesn't need to run in a separate prefix. But, you need to install the game files manually. IIRC the readme has instructions for that

**Comment by vsTerminus** [comment]
I chose to run it in a different prefix because of the dotnet dependency. Maybe it's not an issue anymore but historically installing (or more specifically, _removing_ dotnet) from a prefix could get messy and force you to have to recreate the whole thing. I determined that by running SRS in its own prefix I could easily recreate that prefix at any time without any repercussions should I need to.

You can install SRS into the DCS prefix without needing dotnet48, and then it places all the files you need automatically.

So at the end of the day I have two entries in Lutris: 

- One which launches the app and uses a separate prefix just for SRS
- One which launches the SRS updater and uses the DCS prefix. 

But like I said you could probably just do it all from the DCS prefix if you wanted to.

**Comment by badbabykosh** [comment]
Does anyone have screenshots that show how to do this with lutris?

**Comment by vsTerminus** [comment]
![Screenshot_20241115_154824](https://github.com/user-attachments/assets/06b92934-1088-4872-aaf9-e98ea6bf8839)

It's just two different entries

- One points to the SRS Auto-Updater exe and uses the DCS game prefix
- The other points to the SRS Client Radio exe and uses its own prefix (Where I've installed dotnet48)

I run the auto-updater before I play, and then after I launch DCS I manually launch SRS.

I do the same thing for DCS VR and DCS flatscreen; Different Lutris entries and prefixes for each, allowing me to set different wine/proton runners and different companion apps (eg, the flatscreen prefix has Linuxtrack installed for TrackIR, the VR prefix does not)

![Screenshot_20241115_155611](https://github.com/user-attachments/assets/fd2b3202-0984-440e-8c38-77522a55bd08)

Definitely not _necessary_, but I've found it more convenient this way.

**Comment by vsTerminus** [comment]
Just a little update...

Seems like the steps I mentioned previously are no longer required. I created a fresh prefix and launched with Wine 10.0 and it just works. It actually works _better_ than with dotnet48 installed. I suppose wine-mono has made some improvements.

So no need to run SRS in its own prefix anymore, if you were doing that.

![Image](https://github.com/user-attachments/assets/f50ac658-e41b-4f3d-8eed-f77cd0628520)
